src/Shared/ChangeWaves.cs

```diff
@@ -0,0 +1,204 @@
+
+        public static string DisabledWave
+        {
+            get
+            {
+                if (cachedWave == null)
+                {
+                    cachedWave = Traits.Instance.MSBuildDisableFeaturesFromVersion ?? ChangeWaves.EnableAllFeatures;
+                }
+
+                return cachedWave;
+            }
+            set
+            {
+                cachedWave = value;
+            }
+        }
+
+        private static ChangeWaveConversionState _state;
+        internal static ChangeWaveConversionState ConversionState
+        {
+            get
+            {
+                return _state;
+            }
+            set
+            {
+                // Keep state persistent.
+                if (_state == ChangeWaveConversionState.NotConvertedYet)
+                {
+                    _state = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ensure the the environment variable MSBuildDisableFeaturesFromWave is set to a proper value.
+        /// </summary>
+        /// <returns> String representation of the set change wave. "999.999" if unset or invalid, and clamped if out of bounds. </returns>
+        internal static void ApplyChangeWave()
+        {
+            Version changeWave;
+
+            // If unset, enable all features.
+            if (DisabledWave.Length == 0)
+            {
+                ConversionState = ChangeWaveConversionState.Valid;
+                DisabledWave = ChangeWaves.EnableAllFeatures;
+                return;
+            }
+
+            // If the version is of invalid format, log a warning and enable all features.
+            if (!Version.TryParse(DisabledWave, out changeWave))
+            {
+                ConversionState = ChangeWaveConversionState.InvalidFormat;
+                DisabledWave = ChangeWaves.EnableAllFeatures;
+                return;
+            }
+            // If the version is 999.999, we're done.
+            else if (changeWave == EnableAllFeaturesAsVersion)
+            {
+                ConversionState = ChangeWaveConversionState.Valid;
+                DisabledWave = changeWave.ToString();
+                return;
+            }
+            // If the version is out of rotation, log a warning and clamp the value.
+            else if (changeWave < LowestWaveAsVersion)
+            {
+                ConversionState = ChangeWaveConversionState.OutOfRotation;
+                DisabledWave = LowestWave;
+                return;
+            }
+            else if (changeWave > HighestWaveAsVersion)
+            {
+                ConversionState = ChangeWaveConversionState.OutOfRotation;
+                DisabledWave = HighestWave;
+                return;
+            }
+
+            // Ensure it's set to an existing version within the current rotation
+            if (!AllWavesAsVersion.Contains(changeWave))
+            {
+                foreach (Version wave in AllWavesAsVersion)
```

Comment:
This can be part of the if/else chain or switch statement, right? Also, why do you have to loop through them? Can't you just set it to whatever it is, since `<` (AreFeaturesEnabled) works the same either way?

Reply:
This loop is to find the next available version, because we only get here if:
1. It is a valid version between the bounds
1. it isn't already in the array
So we need to find the next highest one to clamp it to.